before_after test

efischer19 opened the pull request:
Inspired by http://www.oreills.co.uk/2015/03/01/testing-race-conditions-in-python.html

efischer19 commented on openedx/core/djangoapps/course_groups/tests/test_cohorts.py:
```diff
@@ -635,6 +636,30 @@ def test_add_user_to_cohort(self, mock_tracker):
+        membership1 = CohortMembership(user=course_user, course_user_group=first_cohort)
+        membership2 = CohortMembership(user=course_user, course_user_group=second_cohort)
+
+        with before_after.after('models.CohortMembership._innerRead', membership2.save):
```
This line is incorrect, I keep getting the following error:
```
Traceback (most recent call last):
    openedx/core/djangoapps/course_groups/tests/test_cohorts.py line 649 in add_user_to_cohorts_race_condition
      with before_after.after('models.CohortMembership._innerRead', membership2.save):
    /usr/lib/python2.7/contextlib.py line 17 in __enter__
      return self.gen.next()
    /edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/before_after/__init__.py line 56 in before_after
      original, _ = patcher.get_original()
    /edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/mock.py line 1227 in get_original
      target = self.getter()
    /edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/mock.py line 1414 in <lambda>
      getter = lambda: _importer(target)
    /edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/mock.py line 1098 in _importer
      thing = __import__(import_path)
   ImportError: No module named models
```
Testeng (@benpatterson @jzoldak), is this something you could help me get figured out? Or resident python expert @nedbat perhaps? I'm trying to get a patch-able reference to the `_innerRead` method in `CohortMembership`.

benpatterson replied:
@efischer19 I don't suppose you've experimented with fuller references, e.g., `openedx.core.djangoapps.course_groups.models.CohortMembership._innerRead`? Sometimes with patching, I start from there and work my way downwards

benpatterson replied:
You could also try pre-pending two dots in front of 'models' like in your import statements above.

efischer19 replied:
Using the full reference got it there, thanks!

efischer19 commented on requirements/edx/base.txt:
```diff
@@ -6,6 +6,7 @@
 
 beautifulsoup4==4.1.3
 beautifulsoup==3.2.1
+before_after==0.1.3
```
@mollydb, I've been informed by this document that I need to notify you of this change for licensing reasons. This library is licensed under GPLv2 (https://github.com/c-oreills/before_after/blob/master/LICENSE), and I checked in the Open Source chat room before starting this effort.

@edx/devops, `support@edx.org` is also mentioned as needs-to-be-notified above. This library can be found at https://github.com/c-oreills/before_after, it's just sugar on top of the already existing mock library.

jzoldak replied:
@efischer19 can this go in requirements/edx/development.txt instead of base.txt?

efischer19 replied:
I looked into that, but I wasn't sure it would cover our use case here. If this library is needed for unit tests that will be run on Jenkins, will placing it there still keep things working?

jzoldak replied:
Ugh, I just looked and you're right that won't work. The paver command for installing python prereqs doesn't include development.txt.

OK nvm then, carry on.

jzoldak replied:
@efischer19 Oh, actually can you please put it in the "# Used for testing" section instead. At least that way we'll know that it's only for testing and can hopefully untangle those later.

efischer19 replied:
You got it!

efischer19 commented:
I'm attempting to do something similar to http://www.oreills.co.uk/2015/03/01/testing-race-conditions-in-python.html, in order to have a unit test for the functionality added in https://github.com/edx/edx-platform/pull/10295

maxrothman commented:
:+1:

efischer19 commented:
This is now passing locally. @dianakhuang and @jzoldak, would you mind looking this over and thumbs-ing for merge into the main race condition PR?

efischer19 commented:
Also @mollydb, I've outdated the diff originally containing this comment, but I'll wait to get thumbs from you as well on the open source licensing side of things before merging.

As mentioned before, `before_after` is licensed under GPLv2 ()https://github.com/c-oreills/before_after/blob/master/LICENSE, I'm just `requirements.txt`-ing it, and I cleared that with the OS Hipchat room last week.

dianakhuang commented on openedx/core/djangoapps/course_groups/models.py:
```diff
@@ -90,7 +90,7 @@ def save(self, *args, **kwargs):
         max_retries = 5
         for num_attempts in range(max_retries):
-            saved_membership = None
+            saved_membership = self.inner_read()
```
With this change here, do you also need to set `saved_memembership` below? It seems like we're duplicating work.

efischer19 replied:
Good catch - remnants of a bad rebase from this morning. I'll fix that up.

jzoldak commented:
:+1:

dianakhuang commented:
:+1: from me, too.

cahrens commented on openedx/core/djangoapps/course_groups/tests/test_cohorts.py:
```diff
@@ -635,6 +636,37 @@ def test_add_user_to_cohort(self, mock_tracker):
+        first_cohort = CohortFactory(course_id=course.id, name="FirstCohort")
+        second_cohort = CohortFactory(course_id=course.id, name="SecondCohort")
+
+        with before_after.before_after(
```
@efischer19 I'm having trouble understanding this test. Can you add some comments explaining what this is doing?

efischer19 replied:
I've added some more comments, do they help to clarify enough?

cahrens commented:
:+1:

efischer19 commented on openedx/core/djangoapps/course_groups/tests/test_cohorts.py:
```diff
@@ -635,6 +636,45 @@ def test_add_user_to_cohort(self, mock_tracker):
+        # This test confirms that the first add_user_to_cohort call can handle this stale read condition properly.
+        # Proper handling is defined as treating calls as sequential, with write time deciding the order.
+        with before_after.before_after(
+            'django.db.models.Model.save',
```
Note that this is the super(...).save() call that will happen when the `CohortMembership.save()` method tries to create as part of `get_or_create()` https://github.com/edx/edx-platform/blob/efischer/cohort_membership_model/openedx/core/djangoapps/course_groups/models.py#L99

efischer19 commented:
Reviewers - I made a slight change to make this test work with the recent changes made in the parent PR. The only differences between what you previous reviewed and this are:
  * No more `inner_read()` splitting inside the main `save()` method. Since I switched to `select_for_update()`, which uses locks to force an atomic read-then-write, it wouldn't make sense to try and insert a racing `save()` call inside that.
  * Changed the method being targeted by before_after to `django.db.models.Model.save`. This is the new easiest point at which to insert the racing `save()` call.

The test passes locally, unless there are any complaints I'll go ahead and merge once Jenkins tests are passing.

efischer19 commented on openedx/core/djangoapps/course_groups/tests/test_cohorts.py:
```diff
@@ -635,6 +636,45 @@ def test_add_user_to_cohort(self, mock_tracker):
+        second_cohort = CohortFactory(course_id=course.id, name="SecondCohort")
+
+        # This before_after contextmanager allows for reliable reproduction of a race condition.
+        # It will break after the first instance of inner_read, and then run add_user_to_cohort again.
```
Out-of-date documentation, I'll fix this.

efischer19 commented:
jenkins run lettuce
